### PR TITLE
Refactor NPM code in preparation for adding support for PNPM

### PR DESCRIPTION
**Pull request by nickpape-msft**
This PR takes the changes from here: https://github.com/Microsoft/web-build-tools/pull/427

and removes any PNPM-related code. This is so we can make sure that NPM is not broken before adding PNPM support.

### CODE REVIEW

File: common/config/rush/npm-shrinkwrap.json [hunk 1]
```diff
@@ -23,19 +23,19 @@
       "resolved": "https://registry.npmjs.org/@microsoft/gulp-core-build-typescript/-/gulp-core-build-typescript-4.2.10.tgz"
     },
     "@microsoft/node-core-library": {
-      "version": "0.3.12",
```
Comment (nickpape-msft) [inline common/config/rush/npm-shrinkwrap.json]: >1 [](start = 22, length = 1)

I am re-running generate right now

**Comment by iclanton** [comment]
> {

Can you revert this as part of this PR?

---
Refers to: common/config/rush/npm-shrinkwrap.json:1 in f690716. [](commit_id = f69071676cc18c3b3028abfa64c1194e8bb0edc8, deletion_comment = False)

File: apps/rush-lib/src/data/RushConfiguration.ts [hunk 2]
```diff
@@ -267,7 +274,7 @@ export default class RushConfiguration {
    * _validateCommonRushConfigFolder() function makes sure that this folder only contains
    * recognized config files.
    */
-  private static _validateCommonRushConfigFolder(commonRushConfigFolder: string): void {
+  private static _validateCommonRushConfigFolder(commonRushConfigFolder: string, packageManager: PackageManager): void {
```
Comment (iclanton) [inline apps/rush-lib/src/data/RushConfiguration.ts]: >, packageManager: PackageManager [](start = 79, length = 32)

Remove this.

File: apps/rush-lib/src/data/RushConfiguration.ts [hunk 3]
```diff
@@ -378,7 +392,7 @@ export default class RushConfiguration {
    * The filename of the NPM shrinkwrap file that is tracked e.g. by Git.  (The "rush install"
    * command uses a temporary copy, whose path is tempShrinkwrapFilename.)
    * This property merely reports the filename; the file itself may not actually exist.
-   * Example: "C:\MyRepo\common\npm-shrinkwrap.json"
+   * Example: "C:\MyRepo\common\npm-shrinkwrap.json" or "C:\MyRepo\common\shrinkwrap.yaml"
```
Comment (iclanton) [inline apps/rush-lib/src/data/RushConfiguration.ts]: > or "C:\MyRepo\common\shrinkwrap.yaml" [](start = 52, length = 38)

Rename things that reference `*.yaml`

File: apps/rush-lib/src/cli/utilities/InstallManager.ts [hunk 4]
```diff
@@ -140,10 +142,10 @@ export default class InstallManager {
   }
 
   /**
-   * If the "npm-local" symlink hasn't been set up yet, this creates it, installing the
+   * If the "(p)npm-local" symlink hasn't been set up yet, this creates it, installing the
```
Comment (iclanton) [inline apps/rush-lib/src/cli/utilities/InstallManager.ts]: >(p) [](start = 13, length = 3)

Revert

File: apps/rush-lib/src/cli/utilities/InstallManager.ts [hunk 5]
```diff
@@ -597,30 +615,36 @@ export default class InstallManager {
     //       is optional or not, but it does not appear to do so. Also, this would result in strange behavior where
     //       people would have different node_modules based on their system.
 
-    const npmInstallArgs: string[] = ['install', '--no-optional'];
-    this.pushConfigurationNpmArgs(npmInstallArgs);
+    const installArgs: string[] = ['install'];
+    if (this._rushConfiguration.packageManager === 'npm') {
+      installArgs.push('--no-optional');
+    }
```
Comment (iclanton) [inline apps/rush-lib/src/cli/utilities/InstallManager.ts]: Revert this section.

File: apps/rush-lib/src/cli/utilities/Package.ts [hunk 6]
```diff
@@ -85,6 +73,8 @@ export default class Package {
    */
   public parent: Package | undefined;
 
+  public packageJson: IRushTempPackageJson | undefined;
```
Comment (iclanton) [inline apps/rush-lib/src/cli/utilities/Package.ts]: >packageJson [](start = 9, length = 11)

Document or revert.

File: apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts [hunk 7]
```diff
@@ -2,113 +2,37 @@
...
-
-interface IShrinkwrapJson {
-  name: string;
-  version: string;
-  dependencies: { [dependency: string]: IShrinkwrapDependencyJson };
-}
-
 /**
- * This class is a parser for NPM's npm-shrinkwrap.json file format.
+ * This class is a parser for both NPM's npm-shrinkwrap.json file formats.
```
Comment (iclanton) [inline apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts]: > both NPM's npm-shrinkwrap.json file formats. [](start = 29, length = 45)

Revert phrasing .

File: apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts [hunk 8]
```diff
@@ -2,113 +2,37 @@
...
-      if (data.charCodeAt(0) === 0xFEFF) {  // strip BOM
-        data = data.slice(1);
-      }
-
-      return new ShrinkwrapFile(JSON.parse(data));
-    } catch (error) {
-      throw new Error(`Error reading "${shrinkwrapJsonFilename}":` + os.EOL + `  ${error.message}`);
-    }
+  public static loadFromFile(packageManager: PackageManager, shrinkwrapFilename: string): ShrinkwrapFile | undefined {
+    return require('./npm/NpmShrinkwrapFile').default.loadFromFile(shrinkwrapFilename);
```
Comment (iclanton) [inline apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts]: Can you try to refactor this to get rid of the circular dependency?

File: apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts [hunk 9]
```diff
@@ -128,18 +52,23 @@ export default class ShrinkwrapFile {
...
-    // Normalize the data
-    if (!this._shrinkwrapJson.version) {
-      this._shrinkwrapJson.version = '';
-    }
-    if (!this._shrinkwrapJson.name) {
-      this._shrinkwrapJson.name = '';
-    }
-    if (!this._shrinkwrapJson.dependencies) {
-      this._shrinkwrapJson.dependencies = { };
+  protected _getTempProjectNames(dependencies: { [key: string]: {} } ): ReadonlyArray<string> {
```
Comment (qz2017) [inline apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts]: >_getTempProjectNames [](start = 12, length = 20)

Prefer move this one to NpmShrinkwrapFile class. If pnpm needs it in the future, move it at that time.

File: apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts [hunk 10]
```diff
@@ -2,113 +2,37 @@
...
-      data = fsx.readFileSync(shrinkwrapJsonFilename).toString();
-      if (data.charCodeAt(0) === 0xFEFF) {  // strip BOM
-        data = data.slice(1);
-      }
-
-      return new ShrinkwrapFile(JSON.parse(data));
-    } catch (error) {
-      throw new Error(`Error reading "${shrinkwrapJsonFilename}":` + os.EOL + `  ${error.message}`);
-    }
+  public static loadFromFile(packageManager: PackageManager, shrinkwrapFilename: string): ShrinkwrapFile | undefined {
```
Comment (qz2017) [inline apps/rush-lib/src/cli/utilities/ShrinkwrapFile.ts]: >loadFromFile [](start = 16, length = 12)

What is the use of this method?

File: apps/rush-lib/src/cli/utilities/npm/NpmPackage.ts [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+   */
+  versionRange: string;
+
+  /**
+   * The kind of dependency
+   */
+  kind: PackageDependencyKind;
+}
+
+export class NpmPackage extends Package {
```
Comment (qz2017) [inline apps/rush-lib/src/cli/utilities/npm/NpmPackage.ts]: >NpmPackage [](start = 13, length = 10)

Can you please add test for this class?

File: apps/rush-lib/src/data/test/ChangeFile.test.ts [hunk 12]
```diff
@@ -11,7 +11,7 @@ import { ChangeType } from '../ChangeManagement';
 
 describe('ChangeFile', () => {
   it('can add a change', () => {
-    const rushFilename: string = path.resolve(__dirname, 'repo', 'rush.json');
+    const rushFilename: string = path.resolve(__dirname, 'repo', 'rush-npm.json');
```
Comment (qz2017) [inline apps/rush-lib/src/data/test/ChangeFile.test.ts]: >rush-npm [](start = 66, length = 8)

why do we have to rename it?

File: apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+      // Example: "project1"
+      const unscopedTempProjectName: string = Utilities.parseScopedPackageName(project.tempProjectName).name;
+
+      // Example: "C:\MyRepo\common\temp\projects\project1
+      const extractedFolder: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName);
+
+      // Example: "C:\MyRepo\common\temp\projects\project1.tgz"
+      const tarballFile: string = path.join(this._rushConfiguration.commonTempFolder,
+        RushConstants.rushTempProjectsFolderName, unscopedTempProjectName + '.tgz');
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts]: This logic should be changed

Comment (nickpape-msft) [reply apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts]: We probably don't need to read the tarball anymore as the package.json is expected to already exist at the end of rush install

---
In reply to: [152137246](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137246) [](ancestors = 152137246)

File: apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+        RushConstants.nodeModulesFolderName, RushConstants.rushTempNpmScope, unscopedTempProjectName);
+
+      commonProjectPackage = NpmPackage.createVirtualTempPackage(packageJsonFilename, installFolderName);
+
+      // remove the extracted tarball contents
+      fsx.removeSync(packageJsonFilename);
+      fsx.removeSync(extractedFolder);
+
+      commonRootPackage.addChild(commonProjectPackage);
+    }
```
Comment (nickpape-msft) [inline apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts]: extract this?

Comment (nickpape-msft) [reply apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts]: Perhaps I will do this in the next PR?

---
In reply to: [152137695](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137695) [](ancestors = 152137695)

File: apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts (hunk 13 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/utilities/npm/NpmLinkManager.ts]: Will change this later as well

---
In reply to: [152137590](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152137590) [](ancestors = 152137590,152137246)

File: apps/rush-lib/src/data/test/ChangeFile.test.ts (hunk 12 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/data/test/ChangeFile.test.ts]: We renamed it (only in these tests) to distinguish the NPM version of the rush.json from the PNPM version (which differ by some fields)

---
In reply to: [152134278](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152134278) [](ancestors = 152134278)

File: apps/rush-lib/src/cli/utilities/npm/NpmPackage.ts (hunk 11 above)
Comment (nickpape-msft) [reply apps/rush-lib/src/cli/utilities/npm/NpmPackage.ts]: Very difficult to do, plus we are planning to remove this code in the near future.

---
In reply to: [152133530](https://github.com/Microsoft/web-build-tools/pull/446#discussion_r152133530) [](ancestors = 152133530)

**Pull request merged**